core/services/workflows/syncer/handler.go

```diff
@@ -334,103 +334,102 @@ func (w workflowName) Hex() string {
 	return hexName
 }
 
-// workflowRegisteredEvent handles the WorkflowRegisteredEvent event type.
+// workflowRegisteredEvent handles the WorkflowRegisteredEvent event type. This method must remain idempotent.
 func (h *eventHandler) workflowRegisteredEvent(
 	ctx context.Context,
 	payload WorkflowRegistryWorkflowRegisteredV1,
 ) error {
 	wfID := hex.EncodeToString(payload.WorkflowID[:])
+	owner := hex.EncodeToString(payload.WorkflowOwner)
 
-	decodedBinary, config, err := h.workflowArtifactsStore.FetchWorkflowArtifacts(ctx, wfID, payload.BinaryURL, payload.ConfigURL)
+	// If there is not a workflow spec in the database, then fetch artifacts and create it
+	spec, err := h.workflowArtifactsStore.GetWorkflowSpec(ctx, owner, payload.WorkflowName)
 	if err != nil {
-		return err
-	}
-
-	// Always fetch secrets from the SecretsURL
-	var secrets []byte
-	if payload.SecretsURL != "" {
-		secrets, err = h.workflowArtifactsStore.GetSecrets(ctx, payload.SecretsURL, payload.WorkflowID, payload.WorkflowOwner)
+		decodedBinary, config, err := h.workflowArtifactsStore.FetchWorkflowArtifacts(ctx, wfID, payload.BinaryURL, payload.ConfigURL)
 		if err != nil {
-			return fmt.Errorf("failed to get secrets: %w", err)
+			return err
 		}
-	}
 
-	// Calculate the hash of the binary and config files
-	hash, err := pkgworkflows.GenerateWorkflowID(payload.WorkflowOwner, payload.WorkflowName, decodedBinary, config, payload.SecretsURL)
-	if err != nil {
-		return fmt.Errorf("failed to generate workflow id: %w", err)
-	}
-
-	// Pre-check: verify that the workflowID matches; if it doesn’t abort and log an error via Beholder.
-	if !bytes.Equal(hash[:], payload.WorkflowID[:]) {
-		return fmt.Errorf("workflowID mismatch: %x != %x", hash, payload.WorkflowID)
-	}
+		// Always fetch secrets from the SecretsURL
+		var secrets []byte
+		if payload.SecretsURL != "" {
```

Comment:
@justinkaseman We could swap this around with the hash + pre-check above so that if we get a workflow that fails this check we aren't constantly fetching the secrets